src/service.js

```diff
@@ -481,14 +573,14 @@ function getServicePromiseOrNullInternal(holder, id) {
   const services = getServices(holder);
   const s = services[id];
   if (s) {
-    const p = s.promise;
-    if (p) {
-      return p;
-    }
-    if (s.obj) {
+    if (s.promise) {
+      return s.promise;
+    } else if (s.obj) {
       return s.promise = Promise.resolve(s.obj);
+    } else if (!s.ctor && !s.factory) {
```

Comment:
Nit: IMO this is more readable if you remove this branch and put the condition into the `assert()`. This conditional implies that `ctor` and `factory` are relevant to the implementation of this method when they aren't really.